### PR TITLE
[4.0] Feature/New Publishing Workflow

### PR DESCRIPTION
**Pull request by bembelimen**
Hello everyone,

with the Google Summer of Code 2017 project, we implemented a new workflow extension in Joomla! 4, which replaces the current publish/unpublish/trashed/archived state. We want to implement a flexible and unlimited workflow which can be extended for other extensions, too.

Presentation at the JWC 2017: https://www.youtube.com/watch?v=5e-HECq44B0
Presentation at JAB 2018: https://youtu.be/z-TfSEggUZM

### Additional Information
https://docs.joomla.org/Publishing_Workflow
https://docs.joomla.org/GSOC_2017_Project_Ideas#Project_VI:_Publishing_Workflow_in_Joomla.21

### Testing Instructions
Install the PR via the Patchtester and reinstall your Joomla! or download the ZIP file (https://github.com/bembelimen/joomla-cms/archive/publishing-workflow.zip) and make a fresh installation. Afterwards go to the Article manager, there you'll find a new menu entry "Workflows", play with it around. You can also use the implemented default workflow.

Workflows will be assigned in the category parameters.

### Images

grafik

grafik

grafik

Comments, tips, support would be very appreciated!

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_content.ini [hunk 1]
```diff
@@ -3,15 +3,17 @@
 ; License GNU General Public License version 2 or later; see LICENSE.txt, see LICENSE.php
 ; Note : All ini files need to be saved as UTF-8
 
+
 COM_CONTENT="Articles"
-COM_CONTENT_ARTICLE_CONTENT="Content"
 COM_CONTENT_ARTICLES_TITLE="Articles"
+COM_CONTENT_ARTICLE_CONTENT="Content"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_content.ini]: please revert this change

File: administrator/language/en-GB/en-GB.com_content.ini [hunk 2]
```diff
@@ -3,15 +3,17 @@
...
 COM_CONTENT_ATTRIBS_ARTICLE_SETTINGS_LABEL="Options"
 COM_CONTENT_ATTRIBS_FIELDSET_LABEL="Options"
 COM_CONTENT_BATCH_OPTIONS="Batch process the selected articles"
 COM_CONTENT_BATCH_TIP="If a category is selected for move/copy, any actions selected will be applied to the copied or moved articles. Otherwise, all actions are applied to the selected articles."
 COM_CONTENT_CHANGE_ARTICLE="Select or Change article"
 COM_CONTENT_CHANGE_ARTICLE_BUTTON="Select/Change"
+COM_CONTENT_CONFIGURATION="Articles: Options"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_content.ini:18]: please revert this change

File: administrator/language/en-GB/en-GB.com_content.ini [hunk 3]
```diff
@@ -20,38 +22,46 @@ COM_CONTENT_CONFIG_CATEGORY_SETTINGS_DESC="These settings apply for Articles Cat
...
 COM_CONTENT_CREATE_ARTICLE_REDIRECTMENU_DESC="Select the page the user will be redirected to after a successful article submission and after cancel (if not set differently below). The default is to redirect to the home page."
 COM_CONTENT_CREATE_ARTICLE_REDIRECTMENU_LABEL="Submission/Cancel Redirect"
-COM_CONTENT_EDIT_ARTICLE="Edit Article"
-COM_CONTENT_EDITORCONFIG_FIELDSET_LABEL="Configure Edit Screen"
+COM_CONTENT_DEFAULT_WORKFLOW="- Use default -"
 COM_CONTENT_EDITING_LAYOUT="Editing Layout"
+COM_CONTENT_EDITORCONFIG_FIELDSET_LABEL="Configure Edit Screen"
+COM_CONTENT_EDIT_ARTICLE="Edit Article"
 COM_CONTENT_ERROR_ALL_LANGUAGE_ASSOCIATED="A content item set to All languages can't be associated. Associations have not been set."
+COM_CONTENT_ERROR_UPDATE_STATE="Can not set state to the item"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_content.ini:38]: ==> Can't set the state of the item

**Comment by brianteeman** [comment]
I will do a PR against your branch next week to make the language files correct english.

Can you please change the qn to quotename etc

File: administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-15.sql [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+(1, 0, 1, 1, 1, 'Unpublish', '', -1, 1),
+(2, 0, 1, 2, 1, 'Publish', '', -1, 2),
+(3, 0, 1, 3, 1, 'Trash', '', -1, 3),
+(4, 0, 1, 4, 1, 'Archive', '', -1, 4);
+
+--
+-- Creating Associations for existing content
+--
+INSERT INTO `#__workflow_associations` (`item_id`, `state_id`, `extension`)
+SELECT `id`, CASE WHEN `state` = -2 THEN 3 WHEN `state` = 0 THEN 1 WHEN `state` = 2 THEN 4 ELSE 2 END, 'com_content' FROM `#__content`;
```
Comment (bembelimen) [inline administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-15.sql:124]: Hello @nikosdion as discussed on JAB in Cologne, I added the migration INSERT. Could you please look over the file and see, if there is something to improve? Perhaps we have to handle the asset_id in a seperate way? Thank you very much!

Comment (nikosdion) [reply administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-15.sql:124]: Hi Benjamin! Sorry for the late reply. I got a gastrointestinal infection right after JaB and I'm still quite awful.

This query is exactly what I had in mind when we talked! :+1:

Comment (bembelimen) [reply administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-15.sql:124]: Thank you!

**Comment by carlitorweb** [comment]
@bembelimen can you fix the conflict files here, for test it?

**Comment by brianteeman** [comment]
I tried to test this and on apply the database changes I got

"1101 BLOB, TEXT, GEOMETRY or JSON column 'description' can't have a default value"

**Comment by brianteeman** [comment]
Having two columns with states (states and 0)  is confusing and probably not needed

chrome_2018-05-24_09-01-43

I would remove the second one (0) and at the same time give the transitions column a name and not just an icon

**Comment by brianteeman** [comment]
The workflows, transitions and states should have "Names" not "Titles"

COM_WORKFLOW_TITLE="Title"
==>
COM_WORKFLOW_TITLE="Name"

**Comment by brianteeman** [comment]
On a new item there is no visible state - this is confusing and unexpected. For me a new article should have a state displayed and this should be the default for that user

chrome_2018-05-24_09-10-32

chrome_2018-05-24_09-07-55

**Comment by brianteeman** [comment]
It is confusing that sometimes you write State and sometimes you wrote Status

**Comment by brianteeman** [comment]
I dont find using the state icon to open the select list to be intuitive. Maybe its just me. Maybe it should either not display the text until the select is displayed OR it should always display the select. A better UX than me can advise ;)

toggle

**Comment by brianteeman** [comment]
Has this been abandoned?

**Comment by bembelimen** [comment]
No? There will be some updates this months.

**Comment by brianteeman** [comment]
It would be great if the merge conflicts could be resolved so that we can test the latest changes

**Comment by bembelimen** [comment]
We're working on it. We have to rebuild the implementation of com_content

**Comment by bembelimen** [comment]
@brianteeman we have here two different things with the same wording:
1. The "Status" which representate the status of the current viewed item (like a banner is published, a workflow itself is unpublished etc.)
2. The "States" which are the new "layer" in the workflow, where you define your workflow states

I would like to have the 2nd one completly renamed into something "unique" which don't leads to confusion ("All states have a status" e.g.)

Is there any good English term which describes the functionality of "States" without confusion with "Status"?

**Comment by brianteeman** [comment]
I agree 100% with your observation about the similarity and confusion of using status and states although they are technically the correct terms

I will take a look at what others do in this case

**Comment by brianteeman** [comment]
I am still getting my head around it but
> The "States" which are the new "layer" in the workflow, where you define your workflow states

Could that be stages?
eg
Stage 1 create content
Stage 2 review content
Stage 3 Publish content

**Comment by wilsonge** [comment]
I'm going to merge this to push this forward as there are issues keeping this in sync. There's clearly several issues still to work out but people can create PRs as they go

**Comment by brianteeman** [comment]
For the record I dont agree with this being merged in its current state and would like to see it reverted and instead of being on a personal repo moved to a joomla-projects repo like all the others and then issues can be worked on there.

This is nowhere near mergeable

**Comment by infograf768** [comment]
This PR has totally broken a new clean 4.0 installation when trying to create articles.
`500 'v40_multi.wa.state_id' isn't in GROUP BY`
```
0 Argument 1 passed to Joomla\Utilities\ArrayHelpergetColumn() must be of the type array, boolean given, called in /Applications/MAMP/htdocs/installmulti/joomla40/administrator/components/com_content/Model/ArticlesModel.php on line 483
```

**Comment by alikon** [comment]
+ no postgresql

**Comment by ghost** [comment]
@infograf768 similar #21474

**Comment by infograf768** [comment]
@franz-wohlkoenig 

As I have done a fully clean install from 4.0-dev, the new tables exist here.
Issue is in code.

**Comment by ghost** [comment]
a nightly didn't work so a clean Install is better?

**Comment by infograf768** [comment]
My issue above is when using multilang.
Creating an issue now

**Comment by infograf768** [comment]
See https://github.com/joomla/joomla-cms/issues/21475

**Comment by brianteeman** [comment]
Sorry but this is way beyond not ready it is not correct #21476

File: administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-15.sql [hunk 5]
```diff
@@ -0,0 +1,124 @@
+--
+-- Table structure for table `#__workflows`
+--
+
+CREATE TABLE IF NOT EXISTS `#__workflows` (
+  `id` int(10) NOT NULL AUTO_INCREMENT,
+  `asset_id` int(10) DEFAULT 0,
+  `published` tinyint(1) NOT NULL DEFAULT 0,
+  `title` varchar(255) NOT NULL,
+  `description` text NOT NULL DEFAULT '',
```
Comment (alikon) [inline administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-15.sql:10]: The MySQL 'TEXT' type cannot have a DEFAULT value (other than NULL)

File: administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-15.sql [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  PRIMARY KEY (`id`),
+  KEY `asset_id` (`asset_id`),
+  KEY `title` (`title`(191)),
+  KEY `extension` (`extension`(191)),
+  KEY `default` (`default`),
+  KEY `created` (`created`),
+  KEY `created_by` (`created_by`),
+  KEY `modified` (`modified`),
+  KEY `modified_by` (`modified_by`)
+) ENGINE=InnoDB COLLATE=utf8mb4_unicode_ci;
```
Comment (alikon) [inline administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-15.sql:27]: shouldn't be `ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 DEFAULT COLLATE=utf8mb4_unicode_ci` ?

**Comment by laoneo** [comment]
Please open new issues or pull requests regarding the new workflow feature. This pr got merged and all comments will likely get forgotten.

**Comment by alikon** [comment]
ok will do, but ......

File: layouts/joomla/edit/global.php [hunk 7]
```diff
@@ -40,6 +40,11 @@
 	'version_note',
 );
 
+if ($displayData->get('item')->id !== null)
```
Comment (roland-d) [inline layouts/joomla/edit/global.php:43]: @bembelimen Why is this change here? This causes a notice undefined in the plugins when you edit them:
image

**Comment by brianteeman** [comment]
Please open new issues or pull requests regarding the new workflow feature. This pr got merged and all comments will likely get forgotten.
